src/main/java/com/aws/greengrass/mqtt/bridge/clients/IoTCoreClient.java

```diff
@@ -109,41 +146,71 @@ private void publishToIotCore(String topic, byte[] payload) {
 
     @Override
     public synchronized void updateSubscriptions(Set<String> topics, @NonNull Consumer<Message> messageHandler) {
-        this.messageHandler = messageHandler;
-        LOGGER.atDebug().kv("topics", topics).log("Subscribing to IoT Core topics");
-
-        Set<String> topicsToRemove = new HashSet<>(subscribedIotCoreTopics);
-        topicsToRemove.removeAll(topics);
-        topicsToRemove.forEach(s -> {
-            try {
-                unsubscribeFromIotCore(s);
-                LOGGER.atDebug().kv(TOPIC, s).log("Unsubscribed from topic");
-                subscribedIotCoreTopics.remove(s);
-            } catch (InterruptedException | ExecutionException | TimeoutException e) {
-                LOGGER.atError().kv(TOPIC, s).setCause(e).log("Unable to unsubscribe");
-                // If we are unable to unsubscribe, leave the topic in the set so that we can try to remove next time.
+        synchronized (subscribeLock) {
+            if (subscribeFuture != null) {
+                subscribeFuture.cancel(true);
             }
-        });
-
-        Set<String> topicsToSubscribe = new HashSet<>(topics);
-        topicsToSubscribe.removeAll(subscribedIotCoreTopics);
-
-        topicsToSubscribe.forEach(s -> {
-            try {
-                subscribeToIotCore(s);
-                LOGGER.atDebug().kv(TOPIC, s).log("Subscribed to topic");
-                subscribedIotCoreTopics.add(s);
-            } catch (InterruptedException | ExecutionException | TimeoutException e) {
-                LOGGER.atError().kv(TOPIC, s).log("Failed to subscribe");
-            }
-        });
+            this.messageHandler = messageHandler;
+            this.toSubscribeIotCoreTopics = new HashSet<>(topics);
+            LOGGER.atDebug().kv("topics", topics).log("Subscribing to IoT Core topics");
+
+            Set<String> topicsToRemove = new HashSet<>(subscribedIotCoreTopics);
+            topicsToRemove.removeAll(topics);
+            topicsToRemove.forEach(s -> {
+                try {
+                    unsubscribeFromIotCore(s);
+                    LOGGER.atDebug().kv(TOPIC, s).log("Unsubscribed from topic");
+                    subscribedIotCoreTopics.remove(s);
+                } catch (InterruptedException | ExecutionException | TimeoutException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)